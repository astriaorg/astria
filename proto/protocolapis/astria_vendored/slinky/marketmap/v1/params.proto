// source: https://github.com/skip-mev/connect/blob/158cde8a4b774ac4eec5c6d1a2c16de6a8c6abb5/proto/slinky/marketmap/v1/params.proto
syntax = "proto3";
package astria_vendored.slinky.marketmap.v1;

option go_package = "github.com/skip-mev/slinky/x/marketmap/types";

// Params defines the parameters for the x/marketmap module.
message Params {
  // MarketAuthorities is the list of authority accounts that are able to
  // control updating the marketmap.
  repeated string market_authorities = 1;

  // Admin is an address that can remove addresses from the MarketAuthorities
  // list. Only governance can add to the MarketAuthorities or change the Admin.
  string admin = 2;
}
