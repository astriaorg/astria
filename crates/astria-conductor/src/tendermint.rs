use std::time::Duration;

use color_eyre::eyre::{
    self,
    WrapErr as _,
};
use reqwest::{
    Client,
    Response as ReqwestResponse,
};
use serde::{
    de::DeserializeOwned,
    Deserialize,
    Serialize,
};

static VALIDATOR_SET_ENDPOINT: &str = "/cosmos/base/tendermint/v1beta1/validatorsets/";

#[derive(Serialize, Debug)]
pub struct EmptyRequest {}

#[derive(Deserialize, Debug)]
pub struct ValidatorSet {
    pub block_height: String,
    pub validators: Vec<Validator>,
}

impl ValidatorSet {
    /// returns the proposer given the current set by ordering the validators by proposer priority.
    /// the validator with the highest proposer priority is the proposer.
    /// TODO: could there ever be two validators with the same priority?
    pub(crate) fn get_proposer(&mut self) -> eyre::Result<Validator> {
        self.validators.sort_by(|v1, v2| {
            v1.proposer_priority
                .parse::<i64>()
                .unwrap()
                .cmp(&v2.proposer_priority.parse::<i64>().unwrap())
        });
        self.validators
            .first()
            .cloned()
            .ok_or_else(|| eyre::eyre!("no proposer found"))
    }
}

#[derive(Deserialize, Debug, Clone)]
pub struct Validator {
    pub address: String,
    pub pub_key: KeyWithType,
    pub voting_power: String,
    pub proposer_priority: String,
}

#[derive(Deserialize, Debug, Clone)]
pub struct KeyWithType {
    #[serde(rename = "@type")]
    pub key_type: String,
    pub key: String,
}

pub struct TendermintClient {
    endpoint: String,
    http_client: Client,
}

impl TendermintClient {
    pub fn new(endpoint: String) -> eyre::Result<Self> {
        let http_client: Client = Client::builder().timeout(Duration::from_secs(5)).build()?;
        Ok(Self {
            endpoint,
            http_client,
        })
    }

    pub async fn get_proposer_address(&self, height: u64) -> eyre::Result<String> {
        let mut validator_set = self.get_validator_set(height).await?;
        let proposer = validator_set.get_proposer()?;
        Ok(proposer.address)
    }

    pub async fn get_validator_set(&self, height: u64) -> eyre::Result<ValidatorSet> {
        let endpoint: String = format!("{}{}{}", self.endpoint, VALIDATOR_SET_ENDPOINT, height);
        self.do_get::<EmptyRequest, ValidatorSet>(endpoint, None)
            .await
            .wrap_err_with(|| format!("failed to get validator set at height `{height}`"))
    }

    async fn do_get<Req: Serialize + Sized, Resp: DeserializeOwned>(
        &self,
        endpoint: String,
        req: Option<Req>,
    ) -> eyre::Result<Resp> {
        let response: ReqwestResponse = self.http_client.get(&endpoint).json(&req).send().await?;
        response
            .error_for_status()
            .wrap_err("server returned error status")?
            .json::<Resp>()
            .await
            .wrap_err("failed reading server response as json")
    }
}
