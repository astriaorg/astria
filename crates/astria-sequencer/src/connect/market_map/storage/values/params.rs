use std::{
    borrow::Cow,
    fmt::{
        self,
        Debug,
        Formatter,
    },
};

use astria_core::{
    connect::market_map::v2::Params as DomainParams,
    primitive::v1::{
        Address,
        ADDRESS_LEN,
    },
};
use astria_eyre::eyre::bail;
use borsh::{
    BorshDeserialize,
    BorshSerialize,
};
use itertools::Itertools as _;
use telemetry::display::base64;

use super::{
    Value,
    ValueImpl,
};

#[derive(BorshSerialize, BorshDeserialize)]
pub(in crate::connect::market_map) struct Params<'a> {
    market_authorities: Vec<Cow<'a, [u8; ADDRESS_LEN]>>,
    admin: Cow<'a, [u8; ADDRESS_LEN]>,
    // NOTE: All addresses in the `market_authorities` and `admin` have the same prefix.
    prefix: Cow<'a, str>,
}

impl<'a> Debug for Params<'a> {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        f.debug_struct("Params")
            .field(
                "market_authorities",
                &self
                    .market_authorities
                    .iter()
                    .map(|address| base64(address.as_slice()))
                    .join(", "),
            )
            .field("admin", &base64(self.admin.as_slice()).to_string())
            .field("prefix", &self.prefix)
            .finish()
    }
}

impl<'a> From<&'a DomainParams> for Params<'a> {
    fn from(params: &'a DomainParams) -> Self {
        Params {
            market_authorities: params
                .market_authorities
                .iter()
                .map(|address| Cow::Borrowed(address.as_bytes()))
                .collect(),
            admin: Cow::Borrowed(params.admin.as_bytes()),
            prefix: Cow::Borrowed(params.admin.prefix()),
        }
    }
}

impl<'a> From<Params<'a>> for DomainParams {
    fn from(params: Params<'a>) -> Self {
        DomainParams::unchecked_from_parts(
            params
                .market_authorities
                .into_iter()
                .map(|address_bytes| {
                    Address::unchecked_from_parts(
                        address_bytes.into_owned(),
                        params.prefix.as_ref(),
                    )
                })
                .collect(),
            Address::unchecked_from_parts(params.admin.into_owned(), params.prefix.as_ref()),
        )
    }
}

impl<'a> From<Params<'a>> for crate::storage::StoredValue<'a> {
    fn from(params: Params<'a>) -> Self {
        crate::storage::StoredValue::ConnectMarketMap(Value(ValueImpl::Params(params)))
    }
}

impl<'a> TryFrom<crate::storage::StoredValue<'a>> for Params<'a> {
    type Error = astria_eyre::eyre::Error;

    fn try_from(value: crate::storage::StoredValue<'a>) -> Result<Self, Self::Error> {
        let crate::storage::StoredValue::ConnectMarketMap(Value(ValueImpl::Params(params))) = value
        else {
            bail!(
                "connect market map stored value type mismatch: expected params, found {value:?}"
            );
        };
        Ok(params)
    }
}
