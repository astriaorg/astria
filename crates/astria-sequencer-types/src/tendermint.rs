// Calculates the `last_commit_hash` given a Tendermint [`Commit`].
//
// It merkleizes the commit and returns the root. The leaves of the merkle tree
// are the protobuf-encoded [`CommitSig`]s; ie. the signatures that the commit consist of.
//
// See https://github.com/cometbft/cometbft/blob/539985efc7d461668ffb46dff88b3f7bb9275e5a/types/block.go#L922
#[must_use]
pub fn calculate_last_commit_hash(commit: &tendermint::block::Commit) -> tendermint::Hash {
    use prost::Message as _;
    use tendermint::{
        crypto,
        merkle,
    };
    use tendermint_proto::types::CommitSig;

    let signatures = commit
        .signatures
        .iter()
        .map(|commit_sig| CommitSig::from(commit_sig.clone()).encode_to_vec())
        .collect::<Vec<_>>();
    tendermint::Hash::Sha256(merkle::simple_hash_from_byte_vectors::<
        crypto::default::Sha256,
    >(&signatures))
}
