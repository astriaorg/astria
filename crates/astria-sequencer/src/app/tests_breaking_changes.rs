//! The tests in this file are snapshot tests that are expected to break
//! if breaking changes are made to the application, in particular,
//! any changes that can affect the state tree and thus the app hash.
//!
//! If these tests break due to snapshot mismatches, you can update the snapshots
//! with `cargo insta review`, but you MUST mark the respective PR as breaking.
//!
//! Note: there are two actions not tested here: `Ics20Withdrawal` and `IbcRelay`.
//! These are due to the extensive setup needed to test them.
//! If changes are made to the execution results of these actions, manual testing is required.
use std::collections::HashMap;

use astria_core::{
    primitive::v1::{
        asset::DEFAULT_NATIVE_ASSET_DENOM,
        Address,
        RollupId,
    },
    protocol::transaction::v1alpha1::{
        action::{
            BridgeLockAction,
            BridgeUnlockAction,
            IbcRelayerChangeAction,
            SequenceAction,
            TransferAction,
        },
        Action,
        TransactionParams,
        UnsignedTransaction,
    },
    sequencerblock::v1alpha1::block::Deposit,
};
use cnidarium::StateDelta;
use penumbra_ibc::params::IBCParameters;
use prost::Message as _;
use tendermint::{
    abci,
    abci::types::CommitInfo,
    block::Round,
    Hash,
    Time,
};

use crate::{
    app::test_utils::{
        address_from_hex_string,
        default_fees,
        default_genesis_accounts,
        get_alice_signing_key_and_address,
        get_bridge_signing_key_and_address,
        initialize_app,
        initialize_app_with_storage,
        BOB_ADDRESS,
        CAROL_ADDRESS,
    },
    asset::get_native_asset,
    bridge::state_ext::StateWriteExt as _,
    genesis::GenesisState,
    proposal::commitment::generate_rollup_datas_commitment,
};

#[tokio::test]
async fn app_genesis_snapshot() {
    let app = initialize_app(None, vec![]).await;
    insta::assert_json_snapshot!(app.app_hash.as_bytes());
}

#[tokio::test]
async fn app_finalize_block_snapshot() {
    let (alice_signing_key, _) = get_alice_signing_key_and_address();
    let (mut app, storage) = initialize_app_with_storage(None, vec![]).await;

    let bridge_address = Address::from([99; 20]);
    let rollup_id = RollupId::from_unhashed_bytes(b"testchainid");
    let asset_id = get_native_asset().id();

    let mut state_tx = StateDelta::new(app.state.clone());
    state_tx.put_bridge_account_rollup_id(&bridge_address, &rollup_id);
    state_tx
        .put_bridge_account_asset_id(&bridge_address, &asset_id)
        .unwrap();
    app.apply(state_tx);

    // the state changes must be committed, as `finalize_block` will execute the
    // changes on the latest snapshot, not the app's `StateDelta`.
    app.prepare_commit(storage.clone()).await.unwrap();
    app.commit(storage.clone()).await;

    let amount = 100;
    let lock_action = BridgeLockAction {
        to: bridge_address,
        amount,
        asset_id,
        fee_asset_id: asset_id,
        destination_chain_address: "nootwashere".to_string(),
    };
    let sequence_action = SequenceAction {
        rollup_id,
        data: b"hello world".to_vec(),
        fee_asset_id: asset_id,
    };
    let tx = UnsignedTransaction {
        params: TransactionParams {
            nonce: 0,
            chain_id: "test".to_string(),
        },
        actions: vec![lock_action.into(), sequence_action.into()],
    };

    let signed_tx = tx.into_signed(&alice_signing_key);

    let expected_deposit = Deposit::new(
        bridge_address,
        rollup_id,
        amount,
        asset_id,
        "nootwashere".to_string(),
    );
    let deposits = HashMap::from_iter(vec![(rollup_id, vec![expected_deposit.clone()])]);
    let commitments = generate_rollup_datas_commitment(&[signed_tx.clone()], deposits.clone());

    let timestamp = Time::now();
    let block_hash = Hash::try_from([99u8; 32].to_vec()).unwrap();
    let finalize_block = abci::request::FinalizeBlock {
        hash: block_hash,
        height: 1u32.into(),
        time: timestamp,
        next_validators_hash: Hash::default(),
        proposer_address: [0u8; 20].to_vec().try_into().unwrap(),
        txs: commitments.into_transactions(vec![signed_tx.to_raw().encode_to_vec().into()]),
        decided_last_commit: CommitInfo {
            votes: vec![],
            round: Round::default(),
        },
        misbehavior: vec![],
    };

    app.finalize_block(finalize_block.clone(), storage.clone())
        .await
        .unwrap();
    insta::assert_json_snapshot!(app.app_hash.as_bytes());
}

// Note: this tests every action except for `Ics20Withdrawal` and `IbcRelay`.
//
// If new actions are added to the app, they must be added to this test,
// and the respective PR must be marked as breaking.
#[allow(clippy::too_many_lines)]
#[tokio::test]
async fn app_execute_transaction_with_every_action_snapshot() {
    use astria_core::{
        primitive::v1::asset,
        protocol::transaction::v1alpha1::action::{
            FeeAssetChangeAction,
            InitBridgeAccountAction,
            SudoAddressChangeAction,
        },
    };

    let (alice_signing_key, alice_address) = get_alice_signing_key_and_address();
    let (bridge_signing_key, bridge_address) = get_bridge_signing_key_and_address();
    let bob_address = address_from_hex_string(BOB_ADDRESS);
    let carol_address = address_from_hex_string(CAROL_ADDRESS);

    let genesis_state = GenesisState {
        accounts: default_genesis_accounts(),
        authority_sudo_address: alice_address,
        ibc_sudo_address: alice_address,
        ibc_relayer_addresses: vec![],
        native_asset_base_denomination: DEFAULT_NATIVE_ASSET_DENOM.to_string(),
        ibc_params: IBCParameters::default(),
        allowed_fee_assets: vec![DEFAULT_NATIVE_ASSET_DENOM.to_owned().into()],
        fees: default_fees(),
    };
    let (mut app, storage) = initialize_app_with_storage(Some(genesis_state), vec![]).await;

    // setup for ValidatorUpdate action
    let pub_key = tendermint::public_key::PublicKey::from_raw_ed25519(&[1u8; 32]).unwrap();
    let update = tendermint::validator::Update {
        pub_key,
        power: 100u32.into(),
    };

    // setup for BridgeLockAction
    let rollup_id = RollupId::from_unhashed_bytes(b"testchainid");
    let asset_id = get_native_asset().id();
    let mut state_tx = StateDelta::new(app.state.clone());
    state_tx.put_bridge_account_rollup_id(&bridge_address, &rollup_id);
    state_tx
        .put_bridge_account_asset_id(&bridge_address, &asset_id)
        .unwrap();
    app.apply(state_tx);

    let tx = UnsignedTransaction {
        params: TransactionParams {
            nonce: 0,
            chain_id: "test".to_string(),
        },
        actions: vec![
            TransferAction {
                to: bob_address,
                amount: 333_333,
                asset_id: get_native_asset().id(),
                fee_asset_id: get_native_asset().id(),
            }
            .into(),
            SequenceAction {
                rollup_id: RollupId::from_unhashed_bytes(b"testchainid"),
                data: b"hello world".to_vec(),
                fee_asset_id: get_native_asset().id(),
            }
            .into(),
            Action::ValidatorUpdate(update.clone()),
            IbcRelayerChangeAction::Addition(bob_address).into(),
            IbcRelayerChangeAction::Addition(carol_address).into(),
            IbcRelayerChangeAction::Removal(bob_address).into(),
            // TODO: should fee assets be stored in state?
            FeeAssetChangeAction::Addition(asset::Id::from("test-0".to_string())).into(),
            FeeAssetChangeAction::Addition(asset::Id::from("test-1".to_string())).into(),
            FeeAssetChangeAction::Removal(asset::Id::from("test-0".to_string())).into(),
            InitBridgeAccountAction {
                rollup_id: RollupId::from_unhashed_bytes(b"testchainid"),
                asset_id: get_native_asset().id(),
                fee_asset_id: get_native_asset().id(),
            }
            .into(),
            BridgeLockAction {
                to: bridge_address,
                amount: 100,
                asset_id: get_native_asset().id(),
                fee_asset_id: get_native_asset().id(),
                destination_chain_address: "nootwashere".to_string(),
            }
            .into(),
            SudoAddressChangeAction {
                new_address: bob_address,
            }
            .into(),
        ],
    };

    let signed_tx = tx.into_signed(&alice_signing_key);
    app.execute_transaction(signed_tx).await.unwrap();

    // execute BridgeUnlock action
    let tx = UnsignedTransaction {
        params: TransactionParams {
            nonce: 0,
            chain_id: "test".to_string(),
        },
        actions: vec![
            BridgeUnlockAction {
                to: bob_address,
                amount: 10,
                fee_asset_id: get_native_asset().id(),
                memo: vec![0u8; 32],
            }
            .into(),
        ],
    };

    let signed_tx = tx.into_signed(&bridge_signing_key);
    app.execute_transaction(signed_tx).await.unwrap();

    app.prepare_commit(storage.clone()).await.unwrap();
    app.commit(storage.clone()).await;

    insta::assert_json_snapshot!(app.app_hash.as_bytes());
}
