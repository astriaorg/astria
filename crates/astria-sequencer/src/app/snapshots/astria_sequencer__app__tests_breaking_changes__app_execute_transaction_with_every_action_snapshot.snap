---
source: crates/astria-sequencer/src/app/tests_breaking_changes.rs
assertion_line: 274
expression: app.app_hash.as_bytes()
---
[
  30,
  232,
  210,
  6,
  194,
  12,
  154,
  179,
  145,
  105,
  40,
  101,
  30,
  224,
  10,
  195,
  119,
  124,
  207,
  182,
  132,
  175,
  9,
  191,
  104,
  213,
  200,
  146,
  180,
  192,
  229,
  97
]
